src/dataflow/src/source/timestamp.rs

```diff
@@ -199,25 +200,31 @@ impl PartitionTimestamps {
             return None;
         }
 
-        let mut n = self.bindings.len();
+        let mut remaining = self.bindings.len();
+        // We maintain the invariant that `self.bindings[lo].1 <= offset`, but otherwise
+        // increase `lo` to its maximum value. The exstence of a "next" binding at `lo + 1`
+        // determines if we are certain of the timestamp for `offset`.
         let mut lo = 0;
+
         if self.bindings[lo].1 > offset {
             return Some(self.bindings[lo].0);
         }
 
-        while n > 1 {
-            let half = n / 2;
+        while remainaing > 1 {
+            let step = remaining / 2;
 
             // Advance lo if a later element has an offset less than / equal to the one requested.
-            if self.bindings[lo + half].1 <= offset {
-                lo += half;
+            if self.bindings[lo + step].1 <= offset {
+                lo += step;
             }
 
-            n -= half;
+            remaining -= step;
         }
-
+        // If this is not the last binding, we can return the bound time which is
+        // the one found in `self.bindings[lo].0`. The time in `self.bindings[lo+1].0`
+        // is the lower bound on all offsets `self.bindings[lo+1].1` and greater.
         if lo + 1 < self.bindings.len() {
-            Some(self.bindings[lo + 1].0)
+            Some(self.bindings[lo].0)
```

Comment:
This is the only change of consequence; the other stuff is improving comments and variable names.